### ISSUE TITLE
Custom Route Matchers example fails

### ISSUE THREAD

**Issue by gscragg** [sinatra/sinatra#1667]
I am trying to write my own custom route matcher and using the [example in the readme](https://github.com/sinatra/sinatra#custom-route-matchers), I receive the following error: 
`AllButPattern can't be coerced into Mustermann::Pattern (TypeError)`

Here is my example code:

```ruby
require 'sinatra'

get '/index' do
  'Hello world!'
end

class AllButPattern
  Match = Struct.new(:captures)

  def initialize(except)
    @except   = except
    @captures = Match.new([])
  end

  def match(str)
    @captures unless @except === str
  end
end

def all_but(pattern)
  AllButPattern.new(pattern)
end

get all_but("/index") do
  "yep yep yep"
end
```

It would seem as if the API for these matchers has changed with the introduction of Mustermann (4 odd years ago).

**Comment by gscragg** [comment]
I figured out a fix for my problem, but I am not sure its a good one:

```ruby
require 'sinatra'

class AllButPattern

  def initialize(except)
    @except = except
  end

  def to_pattern(**options)
    return self
  end

  def params(route)
    {} unless @except === route
  end
end

def all_but(pattern)
  AllButPattern.new(pattern)
end

get all_but("/index") do
  "yep yep yep"
end
```

**Comment by MatteoPierro** [comment]
this will work as well

```ruby
require 'sinatra'

class AllButPattern < Mustermann::Pattern
  Match = Struct.new(:captures)

  def initialize(except)
    @except   = except
    @captures = Match.new([])
  end

  def ===(str)
    !(@except === str)
  end
end

def all_but(pattern)
  AllButPattern.new(pattern)
end

get all_but("/index") do
  "yep yep yep"
end
```

**Comment by lokcito** [comment]
If your implementation works then this issue should be marked as done, right?

**Comment by dentarg** [comment]
Maybe [the README](https://github.com/sinatra/sinatra/tree/e69b6b9dee7165d3a583fc8a6af10ceee1ea687d#custom-route-matchers) needs an update?

**Comment by bolpin** [comment]
patch can be found at https://github.com/bolpin/sinatra/tree/1667-fix-readme-example-code-errors

**Comment by jkowens** [comment]
Closed by #1769.
